cmd/bosun/sched/check.go

```diff
@@ -200,58 +186,135 @@ func (s *Schedule) RunHistory(r *RunHistory) {
+		r.Events[ak] = &Event{Status: StUnknown}
+	}
+	for _, a := range s.Conf.Alerts {
+		s.CheckAlert(T, r, a)
+	}
+	d := time.Since(start)
+	s.RunHistory(r)
+	<-s.checkRunning
+	return d, nil
+}
+
+var bosunStartupTime = time.Now()
+
+func (s *Schedule) findUnknownAlerts(now time.Time) []expr.AlertKey {
+	keys := []expr.AlertKey{}
+	if time.Now().Sub(bosunStartupTime) < 1*time.Minute {
+		return keys
+	}
+	s.Lock()
+	for ak, st := range s.status {
+		if st.Forgotten {
+			continue
+		}
+		a := s.Conf.Alerts[ak.Name()]
+		t := a.Unknown
+		if t == 0 {
+			t = s.Conf.CheckFrequency * 2
 		}
-		s.Unlock()
-		s.RunHistory(r)
+		if now.Sub(st.Touched) < t {
+			continue
+		}
+		keys = append(keys, ak)
 	}
+	s.Unlock()
+	return keys
 }
 
 func (s *Schedule) CheckAlert(T miniprofiler.Timer, r *RunHistory, a *conf.Alert) {
 	log.Printf("check alert %v start", a.Name)
 	start := time.Now()
-	var warns expr.AlertKeys
-	crits, err := s.CheckExpr(T, r, a, a.Crit, StCritical, nil)
+	var warns, crits expr.AlertKeys
+	d, err := s.executeExpr(T, r, a, a.Depends)
+	var deps expr.ResultSlice
 	if err == nil {
-		warns, _ = s.CheckExpr(T, r, a, a.Warn, StWarning, crits)
+		deps = filterDependencyResults(d)
+		crits, err = s.CheckExpr(T, r, a, a.Crit, StCritical, nil)
+		if err == nil {
+			warns, _ = s.CheckExpr(T, r, a, a.Warn, StWarning, crits)
+		}
 	}
+	depCount, unknownCount := markDependenciesUnevaluated(r.Events, deps, a.Name)
+
 	collect.Put("check.duration", opentsdb.TagSet{"name": a.Name}, time.Since(start).Seconds())
-	log.Printf("check alert %v done (%s): %v crits, %v warns", a.Name, time.Since(start), len(crits), len(warns))
+	log.Printf("check alert %v done (%s): %v crits, %v warns, %v unevaluated, %v unknown", a.Name, time.Since(start), len(crits), len(warns), depCount, unknownCount)
 }
 
-func (s *Schedule) CheckExpr(T miniprofiler.Timer, rh *RunHistory, a *conf.Alert, e *expr.Expr, checkStatus Status, ignore expr.AlertKeys) (alerts expr.AlertKeys, err error) {
-	if e == nil {
-		return
+func filterDependencyResults(results *expr.Results) expr.ResultSlice {
+	// take the results of the dependency expression and filter it to
+	// non-zero tag sets.
+	filtered := expr.ResultSlice{}
+	if results == nil {
+		return filtered
 	}
-	defer func() {
-		if err == nil {
-			return
+	for _, r := range results.Results {
+		var n float64
+		switch v := r.Value.(type) {
+		case expr.Number:
+			n = float64(v)
+		case expr.Scalar:
+			n = float64(v)
 		}
-		collect.Add("check.errs", opentsdb.TagSet{"metric": a.Name}, 1)
-		log.Println(err)
-	}()
-	results, _, err := e.Execute(rh.Context, rh.GraphiteContext, s.Conf.LogstashElasticHosts, rh.Cache, T, rh.Start, 0, a.UnjoinedOK, s.Search, s.Conf.AlertSquelched(a))
+		if !math.IsNaN(n) && n != 0 {
+			filtered = append(filtered, r)
+		}
+	}
+	return filtered
+}
+
+func markDependenciesUnevaluated(events map[expr.AlertKey]*Event, deps expr.ResultSlice, alert string) (depCount, unknownCount int) {
```

Comment:
You changed the var name on the caller. I meant to change it in the named return values here.